node/src/reactor/validator/memory_metrics.rs

```diff
@@ -0,0 +1,274 @@
+    mem_deploy_buffer: IntGauge,
+    /// Estimated heap memory usage of block executor component.
+    mem_block_executor: IntGauge,
+    /// Estimated heap memory usage of block validator component.
+    mem_proto_block_validator: IntGauge,
+    /// Estimated heap memory usage of linear chain component.
+    mem_linear_chain: IntGauge,
+
+    /// Histogram detailing how long it took to measure memory usage.
+    mem_estimator_runtime_s: Histogram,
+
+    /// Instance of registry to unregister from when being dropped.
+    registry: Registry,
+}
+
+impl MemoryMetrics {
+    /// Initializes a new set of memory metrics.
+    pub(super) fn new(registry: Registry) -> Result<Self, prometheus::Error> {
+        let mem_total = IntGauge::new("mem_total", "total memory usage in bytes")?;
+        let mem_metrics = IntGauge::new("mem_metrics", "metrics memory usage in bytes")?;
+        let mem_net = IntGauge::new("mem_net", "net memory usage in bytes")?;
+        let mem_address_gossiper = IntGauge::new(
+            "mem_address_gossiper",
+            "address_gossiper memory usage in bytes",
+        )?;
+        let mem_storage = IntGauge::new("mem_storage", "storage memory usage in bytes")?;
+        let mem_contract_runtime = IntGauge::new(
+            "mem_contract_runtime",
+            "contract_runtime memory usage in bytes",
+        )?;
+        let mem_api_server = IntGauge::new("mem_api_server", "api_server memory usage in bytes")?;
+        let mem_chainspec_loader = IntGauge::new(
+            "mem_chainspec_loader",
+            "chainspec_loader memory usage in bytes",
+        )?;
+        let mem_consensus = IntGauge::new("mem_consensus", "consensus memory usage in bytes")?;
+        let mem_deploy_acceptor = IntGauge::new(
+            "mem_deploy_acceptor",
+            "deploy_acceptor memory usage in bytes",
+        )?;
+        let mem_deploy_fetcher =
+            IntGauge::new("mem_deploy_fetcher", "deploy_fetcher memory usage in bytes")?;
+        let mem_deploy_gossiper = IntGauge::new(
+            "mem_deploy_gossiper",
+            "deploy_gossiper memory usage in bytes",
+        )?;
+        let mem_deploy_buffer =
+            IntGauge::new("mem_deploy_buffer", "deploy_buffer memory usage in bytes")?;
+        let mem_block_executor =
+            IntGauge::new("mem_block_executor", "block_executor memory usage in bytes")?;
+        let mem_proto_block_validator = IntGauge::new(
+            "mem_proto_block_validator",
+            "proto_block_validator memory usage in bytes",
+        )?;
+        let mem_linear_chain =
+            IntGauge::new("mem_linear_chain", "linear_chain memory usage in bytes")?;
+
+        let mem_estimator_runtime_s = Histogram::with_opts(
+            HistogramOpts::new(
+                "mem_estimator_runtime_s",
+                "time taken to estimate memory usage, in seconds",
+            )
+            //  Create buckets from one nanosecond to four seconds.
```

Comment:
Isn't this just up to ~2 seconds?

Reply:
Sounds resonable. I'm changing the lower bound to 4 ns, which should be enough and give us 8 seconds at the top.